cirq-core/cirq/devices/thermal_noise_model.py

```diff
@@ -0,0 +1,251 @@
+
+
+def _lindbladian(left_op: np.ndarray) -> np.ndarray:
+    r"""Superoperator representing a Lindbladian.
+
+    The Lindbladian generated by a single operator A is the superoperator
+
+        $$
+        L(\rho) = A \rho A^\dagger - 0.5 (A^\dagger A \rho + \rho A^\dagger A)
+        $$
+
+    Args:
+        left_op: The operator acting on the left in the Lindbladian (A above).
+
+    Returns:
+        Superoperator corresponding to the Lindbladian.
+    """
+    left_op = np.asarray(left_op)
+    right_op = left_op.conj().T
+    square = right_op @ left_op
+    out = _left_mul(left_op) @ _right_mul(right_op)
+    out -= 0.5 * (_left_mul(square) + _right_mul(square))
+    return out
+
+
+def _decoherence_matrix(
+    cool_rate: float, dephase_rate: float, heat_rate: float = 0.0, dim: int = 2
+) -> np.ndarray:
+    """Construct a rate matrix associated with decay and dephasing.
+
+    The units of the matrix match the units of the rates specified.
+    This matrix can be used to construct a noise channel after rescaling
+    by an idling time (to make it dimensionless).
+
+    Args:
+        cool_rate: Decay rate of the system, usually 1 / T_1
+        dephase_rate: Static dephasing rate of the system, usually 1 / T_phi
+        heat_rate: Heating rate of the system (default 0).
+        dim: Number of energy levels to include (default 2).
+
+    Returns:
+        np.ndarray rate matrix for decay and dephasing.
+    """
+    # Heating (related to a^dag)
+    rate_matrix = np.diag(np.arange(1, dim) * heat_rate, 1).T.astype(float)
+    # Cooling (related to a)
+    rate_matrix += np.diag(np.arange(1, dim) * cool_rate, 1)
+    # Dephasing (related to n=a^dag * a)
+    # We specify i^2 since we take square root to get the Lindblad op later.
+    rate_matrix += np.diag(dephase_rate * np.arange(dim) ** 2)
+    return rate_matrix
+
+
+def _validate_rates(qubit_dims: Dict['cirq.Qid', int], rates: Dict['cirq.Qid', np.ndarray]) -> None:
+    """Check all rate matrices are square and of appropriate dimension.
+
+    We check rates are positive in the class validator.
+    """
+    if set(qubit_dims) != set(rates):
+        raise ValueError('qubits for rates inconsistent with those through qubit_dims')
+    for q in rates:
+        if rates[q].shape != (qubit_dims[q], qubit_dims[q]):
+            raise ValueError(
+                f'Invalid shape for rate matrix: should be ({qubit_dims[q]}, {qubit_dims[q]}), '
+                f'but got {rates[q].shape}'
+            )
+
+
+@dataclass
+class ThermalNoiseModel(devices.NoiseModel):
+    """NoiseModel representing simulated thermalization of a qubit.
+
+    This model is designed for qubits which use energy levels as their states.
+    "Heating" and "cooling" here are used to refer to environmental noise which
+    transitions a qubit to higher or lower energy levels, respectively.
+    """
+
+    def __init__(
+        self,
+        qubit_dims: Dict['cirq.Qid', int],
```

Comment:
Question: If we only support 2, maybe we should just leave this out as an argument ?

Reply:
It would be nice long term to support multi-level qubits, but it's certainly not a requirement for the initial release. The actual reason I kept qubit_dims around was because we do still need the qubits themselves.

You're right about the dimension appearing as a qubit attribute - I've demoted this to `qubits` to simplify things.